pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/EtcdMetadataStore.java

```diff
@@ -0,0 +1,433 @@
+import org.apache.pulsar.metadata.api.extended.CreateOption;
+import org.apache.pulsar.metadata.api.extended.SessionEvent;
+import org.apache.pulsar.metadata.impl.batching.AbstractBatchedMetadataStore;
+import org.apache.pulsar.metadata.impl.batching.MetadataOp;
+import org.apache.pulsar.metadata.impl.batching.OpDelete;
+import org.apache.pulsar.metadata.impl.batching.OpGet;
+import org.apache.pulsar.metadata.impl.batching.OpGetChildren;
+import org.apache.pulsar.metadata.impl.batching.OpPut;
+
+@Slf4j
+public class EtcdMetadataStore extends AbstractBatchedMetadataStore {
+
+    static final String ETCD_SCHEME_IDENTIFIER = "etcd:";
+
+    private final int leaseTTLSeconds;
+    private final Client client;
+    private final KV kv;
+    private volatile long leaseId;
+    private volatile CloseableClient leaseClient;
+    private final EtcdSessionWatcher sessionWatcher;
+
+    public EtcdMetadataStore(String metadataURL, MetadataStoreConfig conf, boolean enableSessionWatcher)
+            throws MetadataStoreException {
+        super(conf);
+
+        this.leaseTTLSeconds = conf.getSessionTimeoutMillis() / 1000;
+        String etcdUrl = metadataURL.replaceFirst(ETCD_SCHEME_IDENTIFIER, "");
+
+        try {
+            this.client = Client.builder().endpoints(etcdUrl).build();
+            this.kv = client.getKVClient();
+            this.client.getWatchClient().watch(ByteSequence.from("\0", StandardCharsets.UTF_8),
+                    WatchOption.newBuilder()
+                            .withPrefix(ByteSequence.from("/", StandardCharsets.UTF_8))
+                            .build(), this::handleWatchResponse);
+            if (enableSessionWatcher) {
+                this.sessionWatcher =
+                        new EtcdSessionWatcher(client, conf.getSessionTimeoutMillis(), this::receivedSessionEvent);
+
+                // Ensure the lease is created when we start
+                this.createLease(false).join();
+            } else {
+                sessionWatcher = null;
+            }
+        } catch (Exception e) {
+            throw new MetadataStoreException(e);
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        super.close();
+
+        if (sessionWatcher != null) {
+            sessionWatcher.close();
+        }
+
+        if (leaseClient != null) {
+            leaseClient.close();
+        }
+
+        if (leaseId != 0) {
+            client.getLeaseClient().revoke(leaseId);
+        }
+
+        kv.close();
+        client.close();
+    }
+
+    private static final GetOption EXISTS_GET_OPTION = GetOption.newBuilder().withCountOnly(true).build();
+    private static final GetOption SINGLE_GET_OPTION = GetOption.newBuilder().withLimit(1).build();
+
+    @Override
+    protected CompletableFuture<Boolean> existsFromStore(String path) {
+        return kv.get(ByteSequence.from(path, StandardCharsets.UTF_8), EXISTS_GET_OPTION)
+                .thenApply(gr -> gr.getCount() == 1);
```

Comment:
Do we need to ensure the async callback is executed in `metadata-store` thread.

Reply:
👍  Fixed it.